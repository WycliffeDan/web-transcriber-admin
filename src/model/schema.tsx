import { KeyMap, Schema, SchemaSettings } from '@orbit/data';

export const keyMap = new KeyMap();

const schemaDefinition: SchemaSettings =  {
    models: {
        book: {
            keys: { remoteId: {} },
            attributes: {
              name: { type: 'string' },
            },
            relationships: {
              type: { type: 'hasOne', model: 'bookType', inverse: 'books' },
              sets: { type: 'hasMany', model: 'set', inverse: 'book' },
            },
        },
        bookType: {
            keys: { remoteId: {} },
            attributes: {
              name: { type: 'string' },
              description: { type: 'string' },
            },
            relationships: {
              books: { type: 'hasMany', model: 'book', inverse: 'bookType' },
            },
        },
        organization: {
            keys: { remoteId: {} },
            attributes: {
              name: { type: 'string' },
              websiteUrl: { type: 'string' },
              logoUrl: { type: 'string' },
              publicByDefault: { type: 'boolean' },
            },
            relationships: {
              owner: { type: 'hasOne', model: 'user', inverse: 'ownedOrganizations' },
              userMemberships: { type: 'hasMany', model: 'user', inverse: 'organization' },
              users: { type: 'hasMany', model: 'user', inverse: 'organizations' },
              projects: { type: 'hasMany', model: 'project', inverse: 'organization' },
              userRoles: { type: 'hasMany', model: 'userRole', inverse: 'organization' },
            }
        },
        project: {
            keys: { remoteId: {} },
            attributes: {
              name: { type: 'string' },
              dateCreated: { type: 'date' },
              dateUpdated: { type: 'date' },
              dateArchived: { type: 'date' },
              language: { type: 'string' },
              description: { type: 'string' },
              allowDownloads: { type: 'boolean' },
              isPublic: { type: 'boolean' },
              // filter keys
              ownerId: { type: 'number' },
              projectTypeId: { type: 'number' },
              organizationId: { type: 'number' },
            },
            relationships: {
              type: { type: 'hasOne', model: 'projectType', inverse: 'projects' },
              owner: { type: 'hasOne', model: 'user', inverse: 'projects' },
              organization: { type: 'hasOne', model: 'organization', inverse: 'projects' },
              // projectIntegrations: { type: 'hasOne', model: 'projectIntegration', inverse: 'projects' },
              projectUsers: { type: 'hasMany', model: 'user', inverse: 'projects' },
              sets: { type: 'hasMany', model: 'set', inverse: 'project' },
              userTasks: { type: 'hasMany', model: 'userTask', inverse: 'project' },
            }
        },
        projectType: {
            keys: { remoteId: {} },
            attributes: {
              name: { type: 'string' },
              description: { type: 'string' },
            },
            relationships: {
              projects: { type: 'hasMany', model: 'project', inverse: 'type' },
            },
        },
        role: {
            keys: { remoteId: {} },
            attributes: {
              roleName: { type: 'string' },
            },
            relationships: {
              userRoles: { type: 'hasMany', model: 'userRole', inverse: 'role' },
            },
        },
        set: {
            keys: { remoteId: {} },
            attributes: {
              name: { type: 'string' },
            },
            relationships: {
              book: { type: 'hasOne', model: 'book', inverse: 'sets' },
              project: { type: 'hasOne', model: 'project', inverse: 'sets' },
            },
        },
        task: {
            keys: { remoteId: {} },
            attributes: {
              reference: { type: 'string' },
              passage: { type: 'string' },
              position: { type: 'number' },
              taskState: { type: 'string' },
              hold: { type: 'number' },
              title: { type: 'string' },
              dateCreated: { type: 'date' },
              dateUpdated: { type: 'date' },
            },
            relationships: {
              userTasks: { type: 'hasMany', model: 'userTask', inverse: 'task' },
              media: { type: 'hasMany', model: 'taskMedia', inverse: 'task' },
              sets: { type: 'hasMany', model: 'set', inverse: 'task' },
            },
        },
        taskMedia: {
            keys: { remoteId: {} },
            attributes: {
              versionNumber: { type: 'number' },
              artifacttype: { type: 'string' },
              eafUrl: { type: 'string' },
              audioUrl: { type: 'string' },
              duration: { type: 'number' },
              contentType: { type: 'string' },
              audioQuality: { type: 'string' },
              textQuality: { type: 'string' },
              transcription: { type: 'string' },
              dateCreated: { type: 'date' },
              dateUpdated: { type: 'date' },
            },
            relationships: {
              task: { type: 'hasOne', model: 'taskMedia', inverse: 'media' },
            },
        },
        user: {
            keys: { remoteId: {} },
            attributes: {
              name: { type: 'string' },
              givenName: { type: 'string' },
              familyName: { type: 'string' },
              auth0Id: { type: 'string' },
              email: { type: 'string' },
              phone: { type: 'string' },
              isLocked: { type: 'boolean' },
              profileVisibility: { type: 'number' },
              emailNotification: { type: 'boolean' },
              timezone: { type: 'string' },
              localization: { type: 'string' },
              decimalSeparator: { type: 'string' },
            },
            relationships: {
              ownedOrganizations: { type: 'hasMany', model: 'organization', inverse: 'owner' },
              organizationMemberships: { type: 'hasMany', model: 'organization', inverse: 'userMemberships' },
              organizations: { type: 'hasMany', model: 'organization', inverse: 'users' },
              assignedTasks: { type: 'hasMany', model: 'userTask', inverse: 'assigned' },
              projects: { type: 'hasMany', model: 'project', inverse: 'owner' },
              userRoles: { type: 'hasMany', model: 'userRole', inverse: 'user' },
            },
        },
        userRole: {
            keys: { remoteId: {} },
            attributes: {
              roleName: { type: 'string' },
            },
            relationships: {
              user: { type: 'hasOne', model: 'user', inverse: 'userRoles' },
              role: { type: 'hasOne', model: 'role', inverse: 'userRoles' },
              organization: { type: 'hasOne', model: 'organization', inverse: 'userRoles' },
            },
        },
        userTask: {
            keys: { remoteId: {} },
            attributes: {
              activityName: { type: 'string' },
              taskState: { type: 'string' },
              comment: { type: 'string' },
              dateCreated: { type: 'date' },
              dateUpdated: { type: 'date' },
            },
            relationships: {
              project: { type: 'hasOne', model: 'project', inverse: 'userTasks' },
              task: { type: 'hasOne', model: 'task', inverse: 'userTasks' },
              assigned: { type: 'hasOne', model: 'user', inverse: 'assignedTasks' },
            },
        },
    }
};

export const schema = new Schema(schemaDefinition);