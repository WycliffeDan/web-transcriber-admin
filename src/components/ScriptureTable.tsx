import React, { useState } from 'react';
import { connect } from 'react-redux';
import { IState, Set, Task, TaskSet, IProjectSettingsStrings } from '../model';
import localStrings from '../selector/localize';
import { withData } from 'react-orbitjs';
import { QueryBuilder } from '@orbit/data';
import { withStyles, WithStyles, Theme } from '@material-ui/core/styles';
import SnackBar from './SnackBar';
import PlanTable from './PlanSheet';

const styles = (theme: Theme) => ({
  container: {
      display: 'flex',
  },
  paper: {
  },
  actions: theme.mixins.gutters({
    paddingBottom: 16,
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'right'
  }),
  button: {
    margin: theme.spacing.unit
  },
  icon: {
    marginLeft: theme.spacing.unit
  },
});

interface IStateProps {
  t: IProjectSettingsStrings;
};

interface IRecordProps {
  sets: Array<Set>;
  tasksets: Array<TaskSet>;
  tasks: Array<Task>;
};

interface IProps extends IStateProps, IRecordProps, WithStyles<typeof styles>{
};
  
export function ScriptureTable(props: IProps) {
    const { classes, t } = props;
    const [message, setMessage] = useState(<></>);
    const [columns, setColumns] = useState([
      {value: 'Section',  readOnly: true, width: 80},
      {value: 'Title',  readOnly: true, width: 280},
      {value: 'Passage', readOnly: true, width: 80},
      {value: 'Book', readOnly: true, width: 100},
      {value: 'Reference', readOnly: true, width: 180},
      {value: 'Description', readOnly: true, width: 280},
    ])
    const [data, setData] = useState([
      [1,"Luke wrote this book about Jesus for Theophilus",'','LUK',"Section 1:1–4",''],
      ['','',1,'LUK',"1:1-4",''],
      [2,"An angel said that John the Baptizer would be born",'','LUK',"Section 1:5–25",''],
      ['','',1,'LUK',"1:5-7",''],
      ['','',2,'LUK',"1:8-10",''],
      ['','',3,'LUK',"1:11-17",''],
      ['','',4,'LUK',"1:18-20",''],
      ['','',5,'LUK',"1:21-25",''],
      [3,"An angel told Mary that Jesus would be born",'','LUK',"Section 1:26–38",''],
      ['','',1,'LUK',"1:26-28",''],
      ['','',2,'LUK',"1:29-34",''],
      ['','',3,'LUK',"1:35-38",''],
      [4,"Mary visited Elizabeth",'','LUK',"Section 1:39–45",''],
      ['','',1,'LUK',"1:39-45",''],
      [5,"Mary praised God",'','LUK',"Section 1:46–56",''],
      ['','',1,'LUK',"1:46-56",''],
      [6,"John the Baptizer was born and received his name",'','LUK',"Section 1:57–66",''],
      ['','',1,'LUK',"1:57-58",''],
      ['','',2,'LUK',"1:59-64",''],
      ['','',3,'LUK',"1:65-66",''],
      [7,"Zechariah prophesied and praised God",'','LUK',"Section 1:67–80",''],
      ['','',1,'LUK',"1:67-80",''],
    ])

    const handleMessageReset = () => { setMessage(<></>) }

    return (
      <div className={classes.container}>
        <PlanTable
          columns={columns}
          rowData={data as any[][]}
        />
        <SnackBar {...props} message={message} reset={handleMessageReset} />
      </div>
    );
}

const mapStateToProps = (state: IState): IStateProps => ({
  t: localStrings(state, {layout: "projectSettings"})
});
    
const mapRecordsToProps = {
  sets: (q: QueryBuilder) => q.findRecords('set'),
  tasks: (q: QueryBuilder) => q.findRecords('task'),
  tasksets: (q: QueryBuilder) => q.findRecords('taskset'),
}

export default withStyles(styles, { withTheme: true })(
  withData(mapRecordsToProps)(
    connect(mapStateToProps)(ScriptureTable) as any
  ) as any
) as any;
      