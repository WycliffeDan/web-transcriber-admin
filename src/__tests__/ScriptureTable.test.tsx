import React from 'react';
import { Router } from 'react-router-dom';
import { setGlobal } from 'reactn';
import { DataProvider } from 'react-orbitjs';
import { Provider } from 'react-redux';
import store from '../store';
import { createMuiTheme, MuiThemeProvider } from '@material-ui/core';
import Store from '@orbit/store';
import { Schema } from '@orbit/data';
import { schema, keyMap } from '../schema';
import history from '../history';
import { render, fireEvent, cleanup, waitForElement } from 'react-testing-library';
import 'jest-dom/extend-expect';
import ScriptureTable from '../components/ScriptureTable';

const theme = createMuiTheme({});
  
const dataStore = new Store({ schema, keyMap });
  
const globals = {
    organization: null,
    project: null,
    plan: null,
    user: null,
    initials: null,
    lang: 'en',
    dataStore: dataStore,
    schema: schema,
    keyMap: keyMap,
}
setGlobal(globals);
  
const tree = (
    <DataProvider dataStore={dataStore}>
      <Provider store={store}>
        <Router history={history}>
          <MuiThemeProvider theme={theme}>
            <ScriptureTable />
          </MuiThemeProvider>
        </Router>
      </Provider>
    </DataProvider>
);

const addOneSection = async () => {
    const plan = {
        type: 'plan',
        attributes: {
            name: 'Genesis',
        },
    } as any;
    (schema as Schema).initializeRecord(plan)
    setGlobal({...globals, plan: plan.id})
    await (dataStore as Store).update(t => t.addRecord(plan));
    const section = {
        type: 'section',
        attributes: {
            sequencenum: 1,
            name: 'Creation',
        }
    } as any;
    (schema as Schema).initializeRecord(section);
    await (dataStore as Store).update(t => t.addRecord(section));
    await (dataStore as Store).update(t => t.replaceRelatedRecord(
        {type: 'section', id: section.id},
        'plan',
        {type: 'plan', id: plan.id}
    ));
    return section.id;
};

const addPassageToSection = async (sectionId: string) => {
    const passage = {
        type: 'passage',
        attributes: {
            sequencenum: 1,
            book: 'GEN',
            reference: '1:1-20',
            position: 0,
            state: '1',
            hold: false,
            title: 'Seven Days'
        }
    } as any;
    (schema as Schema).initializeRecord(passage);
    await (dataStore as Store).update(t => t.addRecord(passage));
    const passageSection = {
        type: 'passagesection',
    } as any;
    (schema as Schema).initializeRecord(passageSection);
    await (dataStore as Store).update(t => t.addRecord(passageSection));
    await (dataStore as Store).update(t => t.replaceRelatedRecord(
        {type: 'passagesection', id: passageSection.id},
        'section',
        {type: 'section', id: sectionId}
    ));
    await (dataStore as Store).update(t => t.replaceRelatedRecord(
        {type: 'passagesection', id: passageSection.id},
        'passage',
        {type: 'passage', id: passage.id}
    ));
    return passage.id;
};

afterEach(cleanup);

test('can render ScriptureTable snapshot', async () => {
    const { getByText, container } = render(tree);
    const TestScriptureTable = await waitForElement(() =>
        getByText(/^Section$/i),
    );
    expect(container.firstChild).toMatchSnapshot();
});

test('ScriptureTable renders on section line', async () => {
    await addOneSection();
    const { getByText, container } = render(tree);
    const TestScriptureTable = await waitForElement(() =>
        getByText(/^Creation$/i),
    );
    expect(getByText(/^Creation$/i)).toHaveTextContent('Creation');
    const body = container.querySelector('tbody');
    expect(body).not.toBeFalsy
    expect(body && body.children.length).toBe(2);
});

test('ScriptureTable renders a section with a passage', async () => {
    const sectionId = await addOneSection();
    await addPassageToSection(sectionId);

    const { getByText, container } = render(tree);
    const TestScriptureTable = await waitForElement(() =>
        getByText(/^Creation$/i),
    );
    expect(getByText(/^Seven Days$/i)).toHaveTextContent('Seven Days');
    const body = container.querySelector('tbody');
    expect(body).not.toBeFalsy
    expect(body && body.children.length).toBe(3);
});

test('ScriptureTable AddPassage button adds a row', async () => {
    const sectionId = await addOneSection();
    await addPassageToSection(sectionId);

    const { getByText, container } = render(tree);
    const TestScriptureTable = await waitForElement(() =>
        getByText(/^Creation$/i),
    );
    fireEvent.click(getByText(/Add Passage/i));
    const body = container.querySelector('tbody');
    expect(body).not.toBeFalsy
    expect(body && body.children.length).toBe(4);
    // sequence number column should be 2
    expect(body && body.children[3].children[3].textContent).toBe('2');
});

test('ScriptureTable AddPassage button adds first row', async () => {
    await addOneSection();

    const { getByText, container } = render(tree);
    const TestScriptureTable = await waitForElement(() =>
        getByText(/^Creation$/i),
    );
    fireEvent.click(getByText(/Add Passage/i));
    const body = container.querySelector('tbody');
    expect(body).not.toBeFalsy
    expect(body && body.children.length).toBe(3);
    // sequence number column should be 1
    expect(body && body.children[2].children[3].textContent).toBe('1');
});

test('ScriptureTable AddSection button adds a row with the next section number', async () => {
    const sectionId = await addOneSection();
    await addPassageToSection(sectionId);

    const { getByText, container } = render(tree);
    const TestScriptureTable = await waitForElement(() =>
        getByText(/^Creation$/i),
    );
    fireEvent.click(getByText(/Add Section/i));
    const body = container.querySelector('tbody');
    expect(body).not.toBeFalsy
    expect(body && body.children.length).toBe(4);
    // sequence number column should be 2
    expect(body && body.children[3].children[1].textContent).toBe('2');
});

test('ScriptureTable AddSection button adds second section when no passages in previous section', async () => {
    await addOneSection();

    const { getByText, container } = render(tree);
    const TestScriptureTable = await waitForElement(() =>
        getByText(/^Creation$/i),
    );
    fireEvent.click(getByText(/Add Section/i));
    const body = container.querySelector('tbody');
    expect(body).not.toBeFalsy
    expect(body && body.children.length).toBe(3);
    // sequence number column should be 2
    expect(body && body.children[2].children[1].textContent).toBe('2');
});

test('ScriptureTable AddSection button adds first section', async () => {
    const { getByText, container } = render(tree);
    const TestScriptureTable = await waitForElement(() =>
        getByText(/^Section$/i),
    );
    fireEvent.click(getByText(/Add Section/i));
    const body = container.querySelector('tbody');
    expect(body).not.toBeFalsy
    expect(body && body.children.length).toBe(2);
    // sequence number column should be 1
    expect(body && body.childNodes[1].childNodes[1].textContent).toBe('1');
});

test('ScriptureTable Delete action with nothing selected', async () => {
    const sectionId = await addOneSection();
    await addPassageToSection(sectionId);

    const { getByText, container } = render(tree);
    const TestScriptureTable = await waitForElement(() =>
        getByText(/^Creation$/i),
    );
    fireEvent.click(getByText(/Action/i));
    const TestScriptureTableActionMenu = await waitForElement(() =>
        getByText(/Delete/i),
    );
    fireEvent.click(getByText(/Delete/i))
    const TestScriptureTableCopyAction = await waitForElement(() =>
        getByText(/Please select row/i),
    );
    expect(true);
});

test('ScriptureTable Select passage row', async () => {
    const sectionId = await addOneSection();
    await addPassageToSection(sectionId);

    const { getByText, getAllByTestId, getByTestId } = render(tree);
    const TestScriptureTable = await waitForElement(() =>
        getByText(/^Creation$/i),
    );
    const box = getAllByTestId('check')[1]
    fireEvent.click(box.querySelector('input') as HTMLElement)
    const TestScriptureTableChecked = await waitForElement(() =>
        getByTestId('checked'),
    );
    expect(true);
});

test('ScriptureTable Delete passage row gives confirmation', async () => {
    const sectionId = await addOneSection();
    await addPassageToSection(sectionId);

    const { getByText, getAllByTestId, getByTestId } = render(tree);
    const TestScriptureTable = await waitForElement(() =>
        getByText(/^Creation$/i),
    );
    const box = getAllByTestId('check')[1]
    fireEvent.click(box.querySelector('input') as HTMLElement)
    const TestScriptureTableChecked = await waitForElement(() =>
        getByTestId('checked'),
    );
    fireEvent.click(getByText(/Action/i));
    const TestScriptureTableActionMenu = await waitForElement(() =>
        getByText(/Delete/i),
    );
    fireEvent.click(getByText(/Delete/i))
    const TestScriptureTableCopyAction = await waitForElement(() =>
        getByText(/Confirmation/i),
    );
    expect(true);
});

test('ScriptureTable Delete passage row', async () => {
    const sectionId = await addOneSection();
    await addPassageToSection(sectionId);

    const { getByText, getAllByTestId, getByTestId, container } = render(tree);
    const TestScriptureTable = await waitForElement(() =>
        getByText(/^Creation$/i),
    );
    const box = getAllByTestId('check')[1]
    fireEvent.click(box.querySelector('input') as HTMLElement)
    const TestScriptureTableChecked = await waitForElement(() =>
        getByTestId('checked'),
    );
    fireEvent.click(getByText(/Action/i));
    const TestScriptureTableActionMenu = await waitForElement(() =>
        getByText(/Delete/i),
    );
    fireEvent.click(getByText(/Delete/i));
    const TestScriptureTableCopyAction = await waitForElement(() =>
        getByText(/Confirmation/i),
    );
    fireEvent.click(getByText(/Yes/i));
    const body = container.querySelector('tbody');
    expect(body && body.children.length).toBe(2);
});

test('ScriptureTable Delete section rows', async () => {
    const sectionId = await addOneSection();
    await addPassageToSection(sectionId);

    const { getByText, getAllByTestId, getByTestId, container } = render(tree);
    const TestScriptureTable = await waitForElement(() =>
        getByText(/^Creation$/i),
    );
    const box = getAllByTestId('check')[0]
    fireEvent.click(box.querySelector('input') as HTMLElement)
    const TestScriptureTableChecked = await waitForElement(() =>
        getAllByTestId('checked'),
    );
    expect(getAllByTestId('checked').length).toBe(2);
    fireEvent.click(getByText(/Action/i));
    const TestScriptureTableActionMenu = await waitForElement(() =>
        getByText(/Delete/i),
    );
    fireEvent.click(getByText(/Delete/i));
    const TestScriptureTableCopyAction = await waitForElement(() =>
        getByText(/Confirmation/i),
    );
    fireEvent.click(getByText(/Yes/i));
    const body = container.querySelector('tbody');
    expect(body && body.children.length).toBe(1); // just the header
});
