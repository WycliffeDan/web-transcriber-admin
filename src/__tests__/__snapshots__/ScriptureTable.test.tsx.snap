// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can render ScriptureTable snapshot 1`] = `
<div
  class="makeStyles-container-1"
>
  <div
    class="makeStyles-container-6"
  >
    <div
      class="makeStyles-paper-7"
    >
      <div
        class="makeStyles-actions-8"
      >
        <button
          aria-label="Action"
          aria-owns="action-menu"
          class="MuiButtonBase-root MuiButton-root makeStyles-button-10 MuiButton-outlined MuiButton-outlinedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Action
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-icon-11"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="Add Section"
          class="MuiButtonBase-root MuiButton-root makeStyles-button-10 MuiButton-outlined MuiButton-outlinedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Add Section
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-icon-11"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="Add Passage"
          class="MuiButtonBase-root MuiButton-root makeStyles-button-10 MuiButton-outlined MuiButton-outlinedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Add Passage
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-icon-11"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <div
          class="makeStyles-grow-9"
        >
          Â 
        </div>
        <button
          aria-label="Save"
          class="MuiButtonBase-root MuiButton-root makeStyles-button-10 MuiButton-contained MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Save
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-icon-11"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <span
        class="data-grid-container"
        tabindex="0"
      >
        <table
          class="data-grid"
        >
          <tbody>
            <tr>
              <td
                class="cell read-only"
              >
                <span
                  class="value-viewer"
                />
              </td>
              <td
                class="cell read-only"
                style="width: 80px;"
              >
                <span
                  class="value-viewer"
                >
                  Section
                </span>
              </td>
              <td
                class="cell read-only"
                style="width: 280px;"
              >
                <span
                  class="value-viewer"
                >
                  Title
                </span>
              </td>
              <td
                class="cell read-only"
                style="width: 80px;"
              >
                <span
                  class="value-viewer"
                >
                  Passage
                </span>
              </td>
              <td
                class="cell read-only"
                style="width: 170px;"
              >
                <span
                  class="value-viewer"
                >
                  Book
                </span>
              </td>
              <td
                class="cell read-only"
                style="width: 180px;"
              >
                <span
                  class="value-viewer"
                >
                  Reference
                </span>
              </td>
              <td
                class="cell read-only"
                style="width: 280px;"
              >
                <span
                  class="value-viewer"
                >
                  Description
                </span>
              </td>
            </tr>
          </tbody>
        </table>
      </span>
    </div>
    <div />
  </div>
</div>
`;
